datajoint/external.py

```diff
@@ -113,29 +90,93 @@ def get(self, blob_hash):
                 pass
 
         if blob is None:
-            try:
-                spec = config[store]
-            except KeyError:
-                raise DataJointError('Store `%s` is not configured' % store)
-
-            try:
-                protocol = spec['protocol']
-            except KeyError:
-                raise DataJointError('Storage {store} config is missing the protocol field'.format(store=store))
-
-            if protocol == 'file':
+            spec = self._get_store_spec(store)
+            if spec['protocol'] == 'file':
                 full_path = os.path.join(spec['location'], self.database, blob_hash)
                 try:
                     with open(full_path, 'rb') as f:
                         blob = f.read()
                 except FileNotFoundError:
-                        raise DataJointError('Lost external blob')
-            elif protocol == 's3':
-                blob = s3.get(self.database, store, blob_hash)
-            else:
-                raise DataJointError('Unknown external storage protocol "%s"' % protocol)
+                    raise DataJointError('Lost external blob %s.' % full_path) from None
+            elif spec['protocol'] == 's3':
+                try:
+                    blob = S3(database=self.database, blob_hash=blob_hash, **spec).get()
+                except TypeError:
+                    raise DataJointError('External store {store} configuration is incomplete.'.format(store=store))
 
             if cache_file:
                 safe_write(cache_file, blob)
 
         return unpack(blob)
+
+    @property
+    def references(self):
+        """
+        :return: generator of referencing table names and their referencing columns
+        """
+        return self.connection.query("""
+        SELECT concat('`', table_schema, '`.`', table_name, '`') as referencing_table, column_name
+        FROM information_schema.key_column_usage
+        WHERE referenced_table_name="{tab}" and referenced_table_schema="{db}"
+        """.format(tab=self.table_name, db=self.database), as_dict=True)
+
+    def delete(self):
+        return self.delete_quick()
+
+    def delete_quick(self):
+        raise DataJointError('The external table does not support delete. Please use delete_garbage instead.')
+
+    def drop(self):
+        """drop the table"""
+        self.drop_quick()
+
+    def drop_quick(self):
+        """drop the external table -- works only when it's empty"""
+        if self:
+            raise DataJointError('Cannot drop a non-empty external table. Please use delete_garabge to clear it.')
+        self.drop_quick()
+
+    def delete_garbage(self):
+        """
+        Delete items that are no longer referenced.
+        This operation is safe to perform at any time.
```

Comment:
Correct me if wrong, but isn't ExternalTable.put called within base relation outside of a transaction in some cases? If so, there conceivably is a minuscule chance of something being added to the backing store (file, s3, etc) and external tracking table before it is referenced in the concrete 'real' relation table, which if deleted here would cause the subsequent update to the user relation to fail; concievably this would leave the backing file 'orphaned' and therefore deletable by clean_store before the issue could be caught and corrected.

not sure if this and the similar race you allude to in the case of clean_store are worth complicating the code with, especially since we are envisioning cache cleaning as a manual maintenance sort of effort and the chances of issues here are very low, but possibly something to note; If i'm correct, this operation is somewhat as risky as clean_store (with the difference that it is only eference rather than real data being purged in the case of clean_store)

it might make sense to flip the external logic to insert the record and then save the blob (rather than save-then-insert ecord) - while this wouldn't prevent this issue as-is, it would allow it to be prevented if we were to lock the entire database during cleanup - writers would then block on the 1st hash insertion and only a 'coherent' database would be cleaned; the other case of failed blob writing would trigger an exception so the external record would remain unreferenced in the main table (and therefore also cleanable)

again, possibly too much given the slim chance; simplest is to document this needs to be done carefully.